File: research/python-lab/src/edgechain_lab/mars/physical.py
"""Physical plausibility scoring for MARS."""

from __future__ import annotations

from dataclasses import dataclass

import numpy as np

from edgechain_lab.fl.metrics import cosine_similarity


@dataclass(frozen=True)
class PhysicalBounds:
    """Odzi-like physical bounds used to generate reference update directions."""

    soil_temp_c: tuple[float, float] = (8.0, 42.0)
    soil_moisture: tuple[float, float] = (0.05, 0.95)
    air_temp_c: tuple[float, float] = (12.0, 38.0)
    humidity: tuple[float, float] = (0.0, 100.0)
    pressure_hpa: tuple[float, float] = (895.0, 925.0)
    lux: tuple[float, float] = (0.0, 110_000.0)


def physical_plausibility_score(
    update_vector: np.ndarray,
    bounds: PhysicalBounds,
    seed: int,
    n_virtual_samples: int = 64,
) -> float:
    """Score whether an update is directionally compatible with plausible bounds.

    This function intentionally accepts no simulator labels, failure codes, or attack metadata.
    """

    if update_vector.size == 0:
        return 0.0

    references = _virtual_reference_updates(
        dimension=update_vector.size,
        bounds=bounds,
        seed=seed,
        n_virtual_samples=n_virtual_samples,
    )
    center = np.mean(references, axis=0)
    direction_score = max(0.0, cosine_similarity(update_vector, center))

    reference_norms = np.linalg.norm(references, axis=1)
    update_norm = float(np.linalg.norm(update_vector))
    mean_norm = float(np.mean(reference_norms))
    std_norm = float(np.std(reference_norms))
    if std_norm == 0.0:
        norm_score = 1.0
    else:
        z_score = abs(update_norm - mean_norm) / std_norm
        norm_score = float(np.clip(1.0 - max(0.0, z_score - 2.0) / 2.0, 0.0, 1.0))

    return float(np.clip(0.7 * direction_score + 0.3 * norm_score, 0.0, 1.0))


def _virtual_reference_updates(
    dimension: int,
    bounds: PhysicalBounds,
    seed: int,
    n_virtual_samples: int,
) -> np.ndarray:
    rng = np.random.default_rng(seed)
    ranges = np.array(
        [
            bounds.soil_temp_c[1] - bounds.soil_temp_c[0],
            bounds.soil_moisture[1] - bounds.soil_moisture[0],
            bounds.air_temp_c[1] - bounds.air_temp_c[0],
            bounds.humidity[1] - bounds.humidity[0],
            bounds.pressure_hpa[1] - bounds.pressure_hpa[0],
            bounds.lux[1] - bounds.lux[0],
        ],
        dtype=float,
    )
    normalized_ranges = ranges / np.linalg.norm(ranges)
    base = np.resize(normalized_ranges, dimension)
    references = []
    for _ in range(n_virtual_samples):
        noise = rng.normal(0.0, 0.05, size=dimension)
        references.append(base + noise)
    return np.asarray(references, dtype=float)

File: packages/mars/src/physical.ts
import { clamp, cosineSimilarity, createSeededNormal, mean, norm, standardDeviation, vectorMean } from './math';
import type { PhysicalBounds } from './types';

export const DEFAULT_PHYSICAL_BOUNDS: PhysicalBounds = {
  soilTempC: [8, 42],
  soilMoisture: [0.05, 0.95],
  airTempC: [12, 38],
  humidity: [0, 100],
  pressureHpa: [895, 925],
  lux: [0, 110_000],
};

export function physicalPlausibilityScore(
  updateVector: number[],
  bounds: PhysicalBounds = DEFAULT_PHYSICAL_BOUNDS,
  seed = 0,
  nVirtualSamples = 64,
): number {
  if (updateVector.length === 0) {
    return 0;
  }

  const references = virtualReferenceUpdates(updateVector.length, bounds, seed, nVirtualSamples);
  const center = vectorMean(references);
  const directionScore = Math.max(0, cosineSimilarity(updateVector, center));

  const referenceNorms = references.map(norm);
  const updateNorm = norm(updateVector);
  const meanNorm = mean(referenceNorms);
  const stdNorm = standardDeviation(referenceNorms);
  const normScore = stdNorm === 0 ? 1 : clamp(1 - Math.max(0, Math.abs(updateNorm - meanNorm) / stdNorm - 2) / 2);

  return clamp(0.7 * directionScore + 0.3 * normScore);
}

function virtualReferenceUpdates(
  dimension: number,
  bounds: PhysicalBounds,
  seed: number,
  nVirtualSamples: number,
): number[][] {
  const ranges = [
    bounds.soilTempC[1] - bounds.soilTempC[0],
    bounds.soilMoisture[1] - bounds.soilMoisture[0],
    bounds.airTempC[1] - bounds.airTempC[0],
    bounds.humidity[1] - bounds.humidity[0],
    bounds.pressureHpa[1] - bounds.pressureHpa[0],
    bounds.lux[1] - bounds.lux[0],
  ];
  const rangeNorm = norm(ranges);
  const normalizedRanges = ranges.map((value) => value / rangeNorm);
  const base = Array.from({ length: dimension }, (_, index) => normalizedRanges[index % normalizedRanges.length]);
  const normal = createSeededNormal(seed);

  return Array.from({ length: nVirtualSamples }, () =>
    base.map((value) => value + normal() * 0.05),
  );
}
